braid/_braid.c

```diff
@@ -2519,20 +2529,109 @@ _braid_FRefineSpace(braid_Core   core,
    {
       *refined_ptr = 0;
    }
-
+    
    /* Reset r_space */
    _braid_CoreElt(core, r_space) = 0;
    return _braid_error_flag;
 }
 
+/*----------------------------------------------------------------------------
+ * Helper routine for periodic problems.  Compute the divisor by which the
+ * number of points must be divisible and also the number of extra needed.
+ *----------------------------------------------------------------------------*/
+
+braid_Int
+_braid_PeriodicCheckNumPoints(braid_Core   core,
+                              braid_Int    npoints,
+                              braid_Int   *divisor_ptr,
+                              braid_Int   *nextra_ptr)
+{
+   braid_Int  divisor, nextra, level, cfactor;
+
+   /* Compute coarsening factor divisor */
+   divisor = 1;
+   nextra  = 0;
+   for (level = 0; level < (_braid_CoreElt(core, max_levels) - 1); level++)
+   {
+      _braid_GetCFactor(core, level, &cfactor);
+      divisor *= cfactor;
+      if ((int)(npoints/divisor) < cfactor)
+      {
+         break;
+      }
+   }
+   if ((npoints % divisor) > 0)
+   {
+      /* Product of coarsening factors (divisor) does not evenly divide npoints */
+
+      /* If we can coarsen one more time, update the divisor to reflect that */
+      if (level < (_braid_CoreElt(core, max_levels) - 1))
```

Comment:
I'm not sure why this if-statement is here?  If this if statement evaluates to true, then that means the above break statement was executed.  Doesn't that mean that we can't coarsen any more?

Reply:
If we are in this if statement, then the product of the coarsening factors does not evenly divide npoints, so we have to do something.  The previous loop exits in one of two ways: 1) we get to a grid size that is too small (i.e., the break statement); 2) we hit max_levels.  This if statement makes sure we didn't hit max_levels.  If we didn't, then when we increase npoints, we will be able to add an additional grid level and the divisor will be larger than currently computed, hence we increase it by the last cfactor we considered in the loop.

Reply:
OK, that makes sense.  Thanks for explaining it!